File: php-api/src/Services/FormAnswerValidator.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

/**
 * Ported from Services/FormAnswerValidator.cs — see that file's own doc comment for the full
 * rationale (validates an AI-Assistant-constructed AnswersJson object against a Form's own
 * FieldsJson schema before FormSubmissionService::create/submit ever sees it; AnswersJson has no
 * server-side validation for the human fill-out path today, since its rendered widgets already
 * enforce required/type/option constraints before an answer can even be typed in — an LLM has no
 * such guardrail, so this is a genuinely NEW validation path scoped to the AI Assistant's
 * submit_form tool only).
 *
 * Mirrors features/form-answers.js's own storage shape exactly: text/textarea -> string, numeric ->
 * number, checkboxGroup -> array of option ids, radio(single) -> bool, radio(mutexGroup)/
 * select(single)/priority -> single option id, radio(multiGroup)/select(multiple) -> array of
 * option ids, datetime -> date string. Option values may be given as either the real option id OR
 * its label (case-insensitive) — normalized to the real id in the cleaned output.
 */
final class FormAnswerValidator
{
    /** Parses fieldsJson into a display-ready field list — [ok, error, fields[]], each field:
     * {id, type, label, helpText, required, multiple, mutex, groupMode, options: [{id,label}]|null}.
     * @return array{0: bool, 1: ?string, 2: array<int, array<string, mixed>>} */
    public static function describeFields(?string $fieldsJson): array
    {
        $fields = self::parseFields($fieldsJson);
        if ($fields === null) {
            return [false, "This form's field definitions could not be read.", []];
        }
        return [true, null, $fields];
    }

    /** Validates $answers against $fieldsJson field-by-field and returns a clean, re-serialized
     * AnswersJson containing only real field ids with normalized option values — any stray key the
     * model supplied that doesn't match a real field id is silently dropped. On failure, the error
     * names the specific field and what's wrong.
     * @return array{0: bool,1: ?string,2: string} [ok, error, answersJson] */
    public static function validate(?string $fieldsJson, array $answers): array
    {
        $fields = self::parseFields($fieldsJson);
        if ($fields === null) {
            return [false, "This form's field definitions could not be read.", '{}'];
        }

        $clean = [];
        foreach ($fields as $field) {
            $provided = array_key_exists($field['id'], $answers) && $answers[$field['id']] !== null;
            $rawValue = $provided ? $answers[$field['id']] : null;
            [$isEmpty, $error, $normalized] = self::validateFieldValue($field, $rawValue);
            if ($error !== null) {
                return [false, $error, '{}'];
            }
            if (!empty($field['required']) && $isEmpty) {
                return [false, "The field \"{$field['label']}\" is required.", '{}'];
            }
            if ($provided && !$isEmpty) {
                $clean[$field['id']] = $normalized;
            }
        }

        return [true, null, json_encode($clean)];
    }

    /** @return ?array<int, array<string, mixed>> */
    private static function parseFields(?string $fieldsJson): ?array
    {
        if ($fieldsJson === null || trim($fieldsJson) === '') {
            return [];
        }
        $decoded = json_decode($fieldsJson, true);
        return is_array($decoded) ? $decoded : null;
    }

    /** @return array{0: bool, 1: ?string, 2: mixed} [isEmpty, error, normalizedValue] */
    private static function validateFieldValue(array $field, mixed $value): array
    {
        if ($value === null) {
            return [true, null, null];
        }
        $type = $field['type'] ?? '';
        $label = $field['label'] ?? $field['id'] ?? '';

        switch ($type) {
            case 'text':
            case 'textarea':
                if (!is_string($value) && !is_numeric($value) && !is_bool($value)) {
                    return [false, "The field \"{$label}\" must be text.", null];
                }
                $s = is_bool($value) ? ($value ? 'true' : 'false') : (string) $value;
                return [trim($s) === '', null, $s];

            case 'numeric':
                if (is_numeric($value)) {
                    return [false, null, $value + 0];
                }
                return [false, "The field \"{$label}\" must be a number.", null];

            case 'datetime':
                $s = is_string($value) ? $value : (string) $value;
                if (trim($s) === '') {
                    return [true, null, null];
                }
                if (strtotime($s) === false) {
                    return [false, "The field \"{$label}\" must be a valid date (e.g. YYYY-MM-DD).", null];
                }
                return [false, null, $s];

            case 'select':
            case 'priority':
                if (!empty($field['multiple'])) {
                    [$ids, $error] = self::resolveOptionArray($field, $value);
                    if ($error !== null) {
                        return [false, $error, null];
                    }
                    return [count($ids) === 0, null, $ids];
                }
                $single = is_string($value) ? $value : (is_scalar($value) ? (string) $value : null);
                if ($single === null || trim($single) === '') {
                    return [true, null, null];
                }
                [$id, $err] = self::matchOption($field, $single);
                if ($err !== null) {
                    return [false, $err, null];
                }
                return [false, null, $id];

            case 'checkboxGroup':
                [$ids, $error] = self::resolveOptionArray($field, $value);
                if ($error !== null) {
                    return [false, $error, null];
                }
                if (!empty($field['mutex']) && count($ids) > 1) {
                    return [false, "Only one option may be selected for \"{$label}\".", null];
                }
                return [count($ids) === 0, null, $ids];

            case 'radio':
                if (($field['groupMode'] ?? null) === 'single') {
                    if (is_bool($value)) {
                        return [$value !== true, null, $value];
                    }
                    if (is_string($value) && in_array(strtolower($value), ['true', 'false'], true)) {
                        $b = strtolower($value) === 'true';
                        return [!$b, null, $b];
                    }
                    return [false, "The field \"{$label}\" must be true or false.", null];
                }
                if (($field['groupMode'] ?? null) === 'multiGroup') {
                    [$ids, $error] = self::resolveOptionArray($field, $value);
                    if ($error !== null) {
                        return [false, $error, null];
                    }
                    return [count($ids) === 0, null, $ids];
                }
                // mutexGroup (the default when groupMode is unset)
                $single = is_string($value) ? $value : (is_scalar($value) ? (string) $value : null);
                if ($single === null || trim($single) === '') {
                    return [true, null, null];
                }
                [$id, $err] = self::matchOption($field, $single);
                if ($err !== null) {
                    return [false, $err, null];
                }
                return [false, null, $id];

            default:
                // An unrecognized field type — pass the raw value through unvalidated rather than
                // blocking the whole submission, same "never let a missing case break the primary
                // flow" posture as elsewhere in this codebase (e.g. PortalQaImageResolver.cs).
                return [false, null, $value];
        }
    }

    /** @return array{0: array<int, string>, 1: ?string} */
    private static function resolveOptionArray(array $field, mixed $value): array
    {
        if (!is_array($value)) {
            return [[], "The field \"" . ($field['label'] ?? $field['id'] ?? '') . "\" must be a list of selected options."];
        }
        $ids = [];
        foreach ($value as $item) {
            $text = is_string($item) ? $item : (is_scalar($item) ? (string) $item : null);
            if ($text === null || trim($text) === '') {
                continue;
            }
            [$id, $err] = self::matchOption($field, $text);
            if ($err !== null) {
                return [[], $err];
            }
            $ids[] = $id;
        }
        return [$ids, null];
    }

    /** @return array{0: ?string, 1: ?string} [id, error] */
    private static function matchOption(array $field, string $valueText): array
    {
        $options = $field['options'] ?? [];
        foreach ($options as $o) {
            if (($o['id'] ?? null) === $valueText) {
                return [$o['id'], null];
            }
        }
        foreach ($options as $o) {
            if (strcasecmp((string) ($o['label'] ?? ''), $valueText) === 0) {
                return [$o['id'], null];
            }
        }
        $label = $field['label'] ?? $field['id'] ?? '';
        $available = count($options) === 0 ? '(no options defined)' : implode(', ', array_map(
            static fn(array $o) => '"' . ($o['label'] ?? '') . '"',
            $options
        ));
        return [null, "\"{$valueText}\" is not a valid option for \"{$label}\". Available: {$available}."];
    }
}

File: api/Enkl.Api/Services/FormAnswerValidator.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Enkl.Api.Services;

/// <summary>
/// Validates an AI-Assistant-constructed AnswersJson object against a Form's own FieldsJson schema
/// before FormSubmissionService.CreateAsync/SubmitAsync ever sees it. AnswersJson is normally "opaque,
/// server-unvalidated" (see Domain/Entities/FormSubmission.cs's own doc comment) — that's fine for the
/// human fill-out UI, since the rendered widgets themselves (features/form-answers.js) already enforce
/// required/type/option constraints before an answer can even be typed in. An LLM constructing the same
/// blob from a schema description has no such guardrail (it can typo a field id, invent an option
/// value, or omit a required field), so this is a genuinely NEW validation path scoped to the AI
/// Assistant's submit_form tool only — not a retrofit of the existing human path, which is a real,
/// separate decision left for later (see AI-ASSISTANT.md).
///
/// Mirrors features/form-answers.js's own storage shape exactly (its module doc comment is the source
/// of truth this was written against): text/textarea -> string, numeric -> number, checkboxGroup ->
/// array of option ids, radio(single) -> bool, radio(mutexGroup)/select(single)/priority -> single
/// option id, radio(multiGroup)/select(multiple) -> array of option ids, datetime -> date string.
/// Option values may be given as either the real option id OR its label (case-insensitive) — normalized
/// to the real id in the cleaned output — same tolerance the existing assigneeName/typeName resolvers
/// already give the model elsewhere in this file, rather than requiring an exact opaque id match.
/// </summary>
public static class FormAnswerValidator
{
    private static readonly JsonSerializerOptions JsonOpts = new() { PropertyNameCaseInsensitive = true };

    private class FieldOption { public string Id { get; set; } = ""; public string Label { get; set; } = ""; }
    private class FieldDef
    {
        public string Id { get; set; } = "";
        public string Type { get; set; } = "";
        public string Label { get; set; } = "";
        public string? HelpText { get; set; }
        public bool Required { get; set; }
        public bool Multiple { get; set; }
        public bool Mutex { get; set; }
        public string? GroupMode { get; set; }
        public List<FieldOption>? Options { get; set; }
    }

    /// <summary>Public projection of FieldDef for callers outside this file (AiAssistantService's
    /// get_form_fields tool) that need to describe a form's schema without duplicating the JSON parsing
    /// this class already does for <see cref="Validate"/> — same field-shape knowledge, one source.</summary>
    public record FieldSummary(
        string Id, string Type, string Label, string? HelpText, bool Required, bool Multiple, bool Mutex,
        string? GroupMode, List<(string Id, string Label)>? Options);

    /// <summary>Parses <paramref name="fieldsJson"/> into a display-ready field list — returns
    /// Ok=false with the same "could not be read" error <see cref="Validate"/> would surface, so a
    /// malformed FieldsJson blob is reported identically regardless of which tool call hit it
    /// first.</summary>
    public static (bool Ok, string? Error, List<FieldSummary> Fields) DescribeFields(string? fieldsJson)
    {
        List<FieldDef> fields;
        try { fields = JsonSerializer.Deserialize<List<FieldDef>>(fieldsJson ?? "[]", JsonOpts) ?? new(); }
        catch (JsonException) { return (false, "This form's field definitions could not be read.", new()); }

        var summaries = fields.Select(f => new FieldSummary(
            f.Id, f.Type, f.Label, f.HelpText, f.Required, f.Multiple, f.Mutex, f.GroupMode,
            f.Options?.Select(o => (o.Id, o.Label)).ToList())).ToList();
        return (true, null, summaries);
    }

    /// <summary>Parses <paramref name="fieldsJson"/>, validates every field's answer in
    /// <paramref name="answers"/>, and returns a clean, re-serialized AnswersJson containing only real
    /// field ids with normalized option values — any stray key the model supplied that doesn't match a
    /// real field id is silently dropped, same defensive-default posture as everywhere else in this
    /// codebase (never let unexpected model-shaped input reach storage verbatim). On failure, Error
    /// names the specific field and what's wrong, so the calling tool can hand it back to the model as
    /// an actionable message it can relay/ask about, rather than a generic failure.</summary>
    public static (bool Ok, string? Error, string AnswersJson) Validate(string? fieldsJson, JsonObject answers)
    {
        List<FieldDef> fields;
        try { fields = JsonSerializer.Deserialize<List<FieldDef>>(fieldsJson ?? "[]", JsonOpts) ?? new(); }
        catch (JsonException) { return (false, "This form's field definitions could not be read.", "{}"); }

        var clean = new JsonObject();
        foreach (var field in fields)
        {
            var provided = answers.TryGetPropertyValue(field.Id, out var rawValue) && rawValue is not null;
            var (isEmpty, error, normalized) = ValidateFieldValue(field, provided ? rawValue : null);
            if (error is not null) return (false, error, "{}");
            if (field.Required && isEmpty) return (false, $"The field \"{field.Label}\" is required.", "{}");
            if (provided && !isEmpty) clean[field.Id] = normalized;
        }

        return (true, null, clean.ToJsonString());
    }

    private static (bool IsEmpty, string? Error, JsonNode? Normalized) ValidateFieldValue(FieldDef field, JsonNode? value)
    {
        if (value is null) return (true, null, null);

        switch (field.Type)
        {
            case "text":
            case "textarea":
            {
                var s = AsString(value);
                if (s is null) return (false, $"The field \"{field.Label}\" must be text.", null);
                return (string.IsNullOrWhiteSpace(s), null, JsonValue.Create(s));
            }
            case "numeric":
            {
                var n = AsNumber(value);
                if (n is null) return (false, $"The field \"{field.Label}\" must be a number.", null);
                return (false, null, JsonValue.Create(n.Value));
            }
            case "datetime":
            {
                var s = AsString(value);
                if (string.IsNullOrWhiteSpace(s)) return (true, null, null);
                if (!DateTime.TryParse(s, out _)) return (false, $"The field \"{field.Label}\" must be a valid date (e.g. YYYY-MM-DD).", null);
                return (false, null, JsonValue.Create(s));
            }
            case "select":
            case "priority":
            {
                if (field.Multiple)
                {
                    var (ids, error) = ResolveOptionArray(field, value);
                    if (error is not null) return (false, error, null);
                    return (ids!.Count == 0, null, ToJsonArray(ids));
                }
                var single = AsString(value);
                if (string.IsNullOrWhiteSpace(single)) return (true, null, null);
                var (id, err) = MatchOption(field, single);
                if (err is not null) return (false, err, null);
                return (false, null, JsonValue.Create(id));
            }
            case "checkboxGroup":
            {
                var (ids, error) = ResolveOptionArray(field, value);
                if (error is not null) return (false, error, null);
                if (field.Mutex && ids!.Count > 1) return (false, $"Only one option may be selected for \"{field.Label}\".", null);
                return (ids!.Count == 0, null, ToJsonArray(ids));
            }
            case "radio" when field.GroupMode == "single":
            {
                var b = AsBool(value);
                if (b is null) return (false, $"The field \"{field.Label}\" must be true or false.", null);
                return (b != true, null, JsonValue.Create(b.Value));
            }
            case "radio" when field.GroupMode == "multiGroup":
            {
                var (ids, error) = ResolveOptionArray(field, value);
                if (error is not null) return (false, error, null);
                return (ids!.Count == 0, null, ToJsonArray(ids));
            }
            case "radio": // mutexGroup (the default when GroupMode is unset)
            {
                var single = AsString(value);
                if (string.IsNullOrWhiteSpace(single)) return (true, null, null);
                var (id, err) = MatchOption(field, single);
                if (err is not null) return (false, err, null);
                return (false, null, JsonValue.Create(id));
            }
            default:
                // An unrecognized field type (a future field type this validator hasn't been taught
                // about yet) — pass the raw value through unvalidated rather than blocking the whole
                // submission, same "never let a missing case break the primary flow" posture as
                // PortalQaImageResolver's own best-effort fallback.
                return (false, null, value.DeepClone());
        }
    }

    private static (List<string>? Ids, string? Error) ResolveOptionArray(FieldDef field, JsonNode value)
    {
        if (value is not JsonArray arr) return (null, $"The field \"{field.Label}\" must be a list of selected options.");
        var ids = new List<string>();
        foreach (var item in arr)
        {
            var text = AsString(item);
            if (string.IsNullOrWhiteSpace(text)) continue;
            var (id, err) = MatchOption(field, text);
            if (err is not null) return (null, err);
            ids.Add(id!);
        }
        return (ids, null);
    }

    private static (string? Id, string? Error) MatchOption(FieldDef field, string valueText)
    {
        var options = field.Options ?? new();
        var byId = options.FirstOrDefault(o => o.Id == valueText);
        if (byId is not null) return (byId.Id, null);
        var byLabel = options.FirstOrDefault(o => string.Equals(o.Label, valueText, StringComparison.OrdinalIgnoreCase));
        if (byLabel is not null) return (byLabel.Id, null);
        var available = options.Count == 0 ? "(no options defined)" : string.Join(", ", options.Select(o => $"\"{o.Label}\""));
        return (null, $"\"{valueText}\" is not a valid option for \"{field.Label}\". Available: {available}.");
    }

    private static JsonArray ToJsonArray(List<string> ids) => new(ids.Select(id => (JsonNode?)JsonValue.Create(id)).ToArray());

    private static string? AsString(JsonNode? node)
    {
        if (node is null) return null;
        try
        {
            var element = node.GetValue<JsonElement>();
            return element.ValueKind switch
            {
                JsonValueKind.String => element.GetString(),
                JsonValueKind.Number => element.ToString(),
                JsonValueKind.True => "true",
                JsonValueKind.False => "false",
                _ => null
            };
        }
        catch (InvalidOperationException) { return node.ToJsonString().Trim('"'); }
    }

    private static double? AsNumber(JsonNode? node)
    {
        if (node is null) return null;
        try
        {
            var element = node.GetValue<JsonElement>();
            if (element.ValueKind == JsonValueKind.Number) return element.GetDouble();
            if (element.ValueKind == JsonValueKind.String && double.TryParse(element.GetString(), out var parsed)) return parsed;
        }
        catch (InvalidOperationException) { /* fall through */ }
        return null;
    }

    private static bool? AsBool(JsonNode? node)
    {
        if (node is null) return null;
        try
        {
            var element = node.GetValue<JsonElement>();
            if (element.ValueKind == JsonValueKind.True) return true;
            if (element.ValueKind == JsonValueKind.False) return false;
            if (element.ValueKind == JsonValueKind.String)
            {
                var s = element.GetString();
                if (string.Equals(s, "true", StringComparison.OrdinalIgnoreCase)) return true;
                if (string.Equals(s, "false", StringComparison.OrdinalIgnoreCase)) return false;
            }
        }
        catch (InvalidOperationException) { /* fall through */ }
        return null;
    }
}
